### ISSUE TITLE
snap auto-refresh breaks cluster

### ISSUE THREAD

**Issue by eug48** [ubuntu/microk8s#1022]
This morning a close-to-production cluster fell over after snap's auto-refresh "feature" failed on 3 of 4 worker nodes - looks like it hanged at the `Copy snap "microk8s" data` step. microk8s could be restarted after aborting the auto-refresh, but this only worked after manually killing snapd.. For a production-ready Kubernetes distribution I really think this is a far from acceptable default.. Perhaps until snapd allows disabling auto-refreshes microk8s scripts could recommend running `sudo snap set system refresh.hold=2050-01-01T15:04:05Z` or similar. Also a kubernetes-native integration with snapd refreshes could be considered (e.g. a prometheus/grafana dashboard/alert) to prompt manual updates - presumably one node at a time to begin with.

Otherwise microk8s is working rather well so thank you very much.

More details about the outage:

```
kubectl get nodes
NAME           STATUS     ROLES    AGE   VERSION
10.aa.aa.aaa   Ready      <none>   38d   v1.17.3
10.aa.aa.aaa   NotReady   <none>   18d   v1.17.2
10.aa.aa.aaa   NotReady   <none>   38d   v1.17.2
10.aa.aa.aaa   NotReady   <none>   18d   v1.17.2
aaa-master     Ready      <none>   59d   v1.17.3
```

microk8s is disabled..
```
root@wk3:/home# snap list
Name      Version    Rev   Tracking  Publisher   Notes
core      16-2.43.3  8689  stable    canonical✓  core
kubectl   1.17.3     1424  1.17      canonical✓  classic
microk8s  v1.17.2    1176  1.17      canonical✓  disabled,classic
```

```
root@wk3:/home# snap changes microk8s
ID   Status  Spawn                Ready  Summary
20   Doing   today at 09:56 AEDT  -      Auto-refresh snap "microk8s"
```

Data copy appears hanged

```
root@wk3:/home# snap tasks --last=auto-refresh
Status  Spawn                Ready                Summary
Done    today at 09:56 AEDT  today at 09:56 AEDT  Ensure prerequisites for "microk8s" are available
Done    today at 09:56 AEDT  today at 09:56 AEDT  Download snap "microk8s" (1254) from channel "1.17/stable"
Done    today at 09:56 AEDT  today at 09:56 AEDT  Fetch and check assertions for snap "microk8s" (1254)
Done    today at 09:56 AEDT  today at 09:56 AEDT  Mount snap "microk8s" (1254)
Done    today at 09:56 AEDT  today at 09:56 AEDT  Run pre-refresh hook of "microk8s" snap if present
Done    today at 09:56 AEDT  today at 09:57 AEDT  Stop snap "microk8s" services
Done    today at 09:56 AEDT  today at 09:57 AEDT  Remove aliases for snap "microk8s"
Done    today at 09:56 AEDT  today at 09:57 AEDT  Make current revision for snap "microk8s" unavailable
Doing   today at 09:56 AEDT  -                    Copy snap "microk8s" data
Do      today at 09:56 AEDT  -                    Setup snap "microk8s" (1254) security profiles
Do      today at 09:56 AEDT  -                    Make snap "microk8s" (1254) available to the system
Do      today at 09:56 AEDT  -                    Automatically connect eligible plugs and slots of snap "microk8s"
Do      today at 09:56 AEDT  -                    Set automatic aliases for snap "microk8s"
Do      today at 09:56 AEDT  -                    Setup snap "microk8s" aliases
Do      today at 09:56 AEDT  -                    Run post-refresh hook of "microk8s" snap if present
Do      today at 09:56 AEDT  -                    Start snap "microk8s" (1254) services
Do      today at 09:56 AEDT  -                    Clean up "microk8s" (1254) install
Do      today at 09:56 AEDT  -                    Run configure hook of "microk8s" snap if present
Do      today at 09:56 AEDT  -                    Run health check of "microk8s" snap
Doing   today at 09:56 AEDT  -                    Consider re-refresh of "microk8s"
```

There doesn't seem to be much to copy anyway:

```
root@wk3 /v/l/snapd# du -sh /var/lib/snapd/ /var/snap/ /snap
527M	/var/lib/snapd/
74G	/var/snap/
2.0G	/snap

root@wk3 /s/microk8s# du -sh /snap/microk8s/*
737M	/snap/microk8s/1176
737M	/snap/microk8s/1254

root@wk3 /s/microk8s# du -sh /var/snap/microk8s/*
232K	/var/snap/microk8s/1176
74G	/var/snap/microk8s/common
```

Starting microk8s fails

```
user@wk3 /s/m/1254> sudo snap start microk8s
error: snap "microk8s" has "auto-refresh" change in progress

root@wk3:/home# snap enable microk8s
error: snap "microk8s" has "auto-refresh" change in progress
```

Fails to abort..

```
root@wk3:/home# snap abort 20
root@wk3:/home# snap changes
ID   Status  Spawn                Ready  Summary
20   Abort   today at 09:56 AEDT  -      Auto-refresh snap "microk8s"

user@wk3 /s/m/1254> sudo snap start microk8s
error: snap "microk8s" has "auto-refresh" change in progress

root@wk3:/home# snap enable microk8s
error: snap "microk8s" has "auto-refresh" change in progress
```

snapd service hangs when trying to stop it...
```
root@wk2 ~# systemctl stop snapd.service
(hangs)
```

have to resort to manually stopping the process

```
killall snapd
```

finally change is undone..

```
root@wk3:/home# snap changes
ID   Status  Spawn                Ready                Summary
20   Undone  today at 09:56 AEDT  today at 10:41 AEDT  Auto-refresh snap "microk8s"

root@wk3:/home# snap tasks --last=auto-refresh
Status  Spawn                Ready                Summary
Done    today at 09:56 AEDT  today at 10:41 AEDT  Ensure prerequisites for "microk8s" are available
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Download snap "microk8s" (1254) from channel "1.17/stable"
Done    today at 09:56 AEDT  today at 10:41 AEDT  Fetch and check assertions for snap "microk8s" (1254)
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Mount snap "microk8s" (1254)
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Run pre-refresh hook of "microk8s" snap if present
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Stop snap "microk8s" services
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Remove aliases for snap "microk8s"
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Make current revision for snap "microk8s" unavailable
Undone  today at 09:56 AEDT  today at 10:41 AEDT  Copy snap "microk8s" data
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Setup snap "microk8s" (1254) security profiles
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Make snap "microk8s" (1254) available to the system
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Automatically connect eligible plugs and slots of snap "microk8s"
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Set automatic aliases for snap "microk8s"
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Setup snap "microk8s" aliases
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Run post-refresh hook of "microk8s" snap if present
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Start snap "microk8s" (1254) services
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Clean up "microk8s" (1254) install
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Run configure hook of "microk8s" snap if present
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Run health check of "microk8s" snap
Hold    today at 09:56 AEDT  today at 10:30 AEDT  Consider re-refresh of "microk8s

root@wk3:/home# snap list
Name      Version    Rev   Tracking  Publisher   Notes
core      16-2.43.3  8689  stable    canonical✓  core
kubectl   1.17.3     1424  1.17      canonical✓  classic
microk8s  v1.17.2    1176  1.17      canonical✓  classic
```

Nothing much in snapd logs except for a polkit error - unsure if related:

```
root@wk3:/home# journalctl -b -u snapd.service

...
Mar 09 06:11:34 wk3 snapd[15182]: autorefresh.go:397: auto-refresh: all snaps are up-to-date
Mar 09 16:11:31 wk3 snapd[15182]: storehelpers.go:436: cannot refresh: snap has no updates available: "core", "kubectl", "microk8s"
Mar 09 16:11:31 wk3 snapd[15182]: autorefresh.go:397: auto-refresh: all snaps are up-to-date
Mar 09 19:06:31 wk3 snapd[15182]: storehelpers.go:436: cannot refresh: snap has no updates available: "core", "kubectl", "microk8s"
Mar 09 19:06:31 wk3 snapd[15182]: autorefresh.go:397: auto-refresh: all snaps are up-to-date
Mar 10 02:51:31 wk3 snapd[15182]: storehelpers.go:436: cannot refresh: snap has no updates available: "core", "kubectl", "microk8s"
Mar 10 02:51:31 wk3 snapd[15182]: autorefresh.go:397: auto-refresh: all snaps are up-to-date
Mar 10 09:56:31 wk3 snapd[15182]: storehelpers.go:436: cannot refresh: snap has no updates available: "core", "kubectl"
Mar 10 10:12:18 wk3 snapd[15182]: daemon.go:208: polkit error: Authorization requires interaction
Mar 10 10:39:24 wk3 systemd[1]: Stopping Snappy daemon...
Mar 10 10:39:24 wk3 snapd[15182]: main.go:155: Exiting on terminated signal.
Mar 10 10:40:54 wk3 systemd[1]: snapd.service: State 'stop-sigterm' timed out. Killing.
Mar 10 10:40:54 wk3 systemd[1]: snapd.service: Killing process 15182 (snapd) with signal SIGKILL.
Mar 10 10:40:54 wk3 systemd[1]: snapd.service: Main process exited, code=killed, status=9/KILL
Mar 10 10:40:54 wk3 systemd[1]: snapd.service: Failed with result 'timeout'.
Mar 10 10:40:54 wk3 systemd[1]: Stopped Snappy daemon.
Mar 10 10:40:54 wk3 systemd[1]: snapd.service: Triggering OnFailure= dependencies.
Mar 10 10:40:54 wk3 systemd[1]: snapd.service: Found left-over process 16729 (sync) in control group while starting unit. Ignoring.
Mar 10 10:40:54 wk3 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Mar 10 10:40:54 wk3 systemd[1]: Starting Snappy daemon...
Mar 10 10:40:54 wk3 snapd[18170]: AppArmor status: apparmor is enabled and all features are available
Mar 10 10:40:54 wk3 snapd[18170]: AppArmor status: apparmor is enabled and all features are available
Mar 10 10:40:54 wk3 snapd[18170]: daemon.go:346: started snapd/2.43.3 (series 16; classic) ubuntu/18.04 (amd64) linux/4.15.0-88-generic.
Mar 10 10:40:54 wk3 snapd[18170]: daemon.go:439: adjusting startup timeout by 45s (pessimistic estimate of 30s plus 5s per snap)
Mar 10 10:40:54 wk3 systemd[1]: Started Snappy daemon.
```

**Comment by eug48** [comment]
I've left one worker node in the stuck state in case that's useful for troubleshooting, and have now come across a well-known issue in that pods running on that `NotReady` node are stuck as `Terminating`. With prometheus this is a problem because the StatefulSet will not start another instance until that pod is manually force-deleted. Just mentioning here as this is another reason why I think microk8s is not ready for auto-refreshes.

**Comment by ktsakalozos** [comment]
Thank you for reporting this @eug48. I opened an issue/topic with the snap team at [1].

One note here is that you cannot hold snap refreshes forever (`sudo snap set system refresh.hold=2050-01-01T15:04:05Z` does not work). You can defer refreshes for up to 90 days, I think. If you want to block refreshes you need to setup a snap store proxy [2]

[1] https://forum.snapcraft.io/t/snap-refresh-breaks-microk8s-cluster/15906
[2] https://docs.ubuntu.com/snap-store-proxy/en/

**Comment by eug48** [comment]
Thanks very much for raising that @ktsakalozos and correcting my incorrect assumption that `refresh.hold` would work long-term. I was misled by the lead sentence "Use refresh.hold to delay snap refreshes until a defined time and date." in snap docs ([1]) and by the command running without an error/warning.. Another lesson to not skim documentation..

[1]: https://snapcraft.io/docs/keeping-snaps-up-to-date#heading--refresh-hold

**Comment by mvo5** [comment]
@eug48 sorry for the trouble and thanks for the report. The fact that it hangs during the copy-data phase is curious. I think you mentioned you have one node in the bad state? 

It looks like the data in /var/snap/microk8s/ did not even got started to get copied, i.e. the new snaps data dir did not even get created, is this correct?

**Comment by eug48** [comment]
@mvo5 yes, /snap/microk8s/1254 got created but in /var/snap/microk8s/ there is only 1176.

Upon further investigation I've probably found the cause. I've been trying out rook-ceph and there is still a volume mounted with it:

```
/dev/rbd0 on /var/snap/microk8s/common/var/lib/kubelet/pods/4dbf852e-f740-4a9f-b72d-de1b50120983/volumes/kubernetes.io~csi/pvc-149ca422-8e37-48f4-b087-98cd31d06c43/mount type ext4 (rw,relatime,stripe=1024,data=ordered)
```

However trying to `ls` some sub-directories within it hangs forever and `dmesg` is full of errors like `libceph: mon0 10.152.183.195:6789 socket error on read`. So ceph is failing to connect to its service, which is running inside the cluster, but flanneld has been stopped.

`sync` also hangs forever and sure enough it looks like snapd has launched a `sync` on which it is presumably waiting.

So this is already a complex and therefore rather brittle set-up, and I think having snap auto-refreshes added to the mix makes failure much more likely. Having an option for it to be turned off permanently so that users can upgrade manually and fix these kinds of problems would be great for production use.

**Comment by ShadowJonathan** [comment]
For anyone reading this with the same issue, `snapd` currently doesnt allow **any indefinite auto-update disabling**, other than this suggestion through a forum thread with a bigger umbrella issue discussion about this; https://forum.snapcraft.io/t/disabling-automatic-refresh-for-snap-from-store/707/268

TL;DR: `snap download foo ; snap install foo.snap --dangerous`, replace `foo` with application in question.

Personally, the fact that `snapd` doesn't allow more comprehensive and cooperative options like (web)hooks on snap refresh, and have an external system handle manual refreshes one-by-one (for example by draining a node first before refreshing it, running some canary checks on it, then doing the same one-by-one for the rest, reverting entirely on error), would be of great help and would even be still within the ethos of keeping snaps up-to-date, but the system to trust sysadmins to make it so needs to be there.

**Comment by skobow** [comment]
I also have problems with my microk8s cluster that my be related to this issue. I am experiencing regular service failure that start almost exactly at 2am in a not yet discovered interval (some days). The exact service is a VerneMQ MQTT server that is using a MariaDB for authentication. The result is that authentication does not work after that (unknown) event happened. This event could relate to snap activities as I discovered some micro8ks restarting by snap around that time.
I also had this behavior with other services. My assumption is that the failure may be related to persistent network connections that fail on k8s level and the applications do not notice the failure. After rescheduling the corresponding pods everything works fine again.

I also would like to disable auto refresh for microk8s to further investigate the problem and to proof my assumption.
Does anyone have any other ideas?

**Comment by ktsakalozos** [comment]
Hi @skobow, is it possible you were following the `latest/edge` channel? What do you get from `snap list | grep microk8s`?

**Comment by skobow** [comment]
Hi @ktsakalozos, I am using 1.19/stable channel which currently installs v1.19.2

**Comment by ktsakalozos** [comment]
Nothing got released on `1.19/stable`. You could attach the `microk8s.inspect` tarball so we can take a look.

**Comment by skobow** [comment]
Find the tarball attached.
The reason for my assumtion is the output of `snap changes` that shows:

`ID   Status  Spawn                Ready                Summary`
`166  Done    today at 02:51 CEST  today at 02:51 CEST  Running service command for snap "microk8s"`
`167  Done    today at 02:51 CEST  today at 02:52 CEST  Running service command for snap "microk8s"`
`168  Done    today at 02:52 CEST  today at 02:52 CEST  Running service command for snap "microk8s"`
`169  Done    today at 02:52 CEST  today at 02:52 CEST  Running service command for snap "microk8s"`
`170  Done    today at 02:52 CEST  today at 02:52 CEST  Running service command for snap "microk8s"`
`171  Done    today at 04:35 CEST  today at 04:35 CEST  Auto-refresh snaps "core", "snapd"`
`172  Done    today at 04:51 CEST  today at 04:51 CEST  Running service command for snap "microk8s"`
`173  Done    today at 04:51 CEST  today at 04:52 CEST  Running service command for snap "microk8s"`
`174  Done    today at 04:52 CEST  today at 04:52 CEST  Running service command for snap "microk8s"`
`175  Done    today at 04:52 CEST  today at 04:52 CEST  Running service command for snap "microk8s"`
`176  Done    today at 04:52 CEST  today at 04:52 CEST  Running service command for snap "microk8s"`
`177  Done    today at 05:08 CEST  today at 05:08 CEST  Running service command for snap "microk8s"`
`178  Done    today at 05:08 CEST  today at 05:08 CEST  Running service command for snap "microk8s"`
`179  Done    today at 05:08 CEST  today at 05:08 CEST  Running service command for snap "microk8s"`
`180  Done    today at 05:08 CEST  today at 05:08 CEST  Running service command for snap "microk8s"`
`181  Done    today at 05:08 CEST  today at 05:08 CEST  Running service command for snap "microk8s"`
`182  Error   today at 10:27 CEST  today at 10:27 CEST  Change configuration of "core" snap`
`183  Done    today at 10:29 CEST  today at 10:29 CEST  Change configuration of "core" snap`

`snap change 166`then shows:

`Status  Spawn                Ready                Summary`
`Done    today at 02:51 CEST  today at 02:51 CEST  restart of [microk8s.daemon-etcd]`

The time stamps fit for the service stop working. Even though there might not be any updates something happens anyway. Could that be related?
[inspection-report-20201014_103630.tar.gz](https://github.com/ubuntu/microk8s/files/5376826/inspection-report-20201014_103630.tar.gz)

**Comment by skobow** [comment]
Hi! Fyi: exactly the same happened tonight at the same time. @ktsakalozos what are these service commands and why are they run?

**Comment by ktsakalozos** [comment]
I am not sure why snapd decides to restart MicroK8s. Could you attach the snapd log `journalctl -u snapd -n 3000`. If we do not see anything there we may need to ask over at https://forum.snapcraft.io/

**Comment by skobow** [comment]
@ktsakalozos find the log attached!

[snapd.log](https://github.com/ubuntu/microk8s/files/5393836/snapd.log)

**Comment by skobow** [comment]
@ktsakalozos Any news on this topic?

**Comment by ktsakalozos** [comment]
@skobow in the snapd.log I see these failures:
```
Oct 16 13:04:32 k8s-master snapd[803]: storehelpers.go:551: cannot refresh: snap has no updates available: "core", "core18", "lxd", "microk8s", "snapd"
Oct 16 13:04:32 k8s-master snapd[803]: stateengine.go:150: state ensure error: cannot sections: got unexpected HTTP status code 403 via GET to "https://api.snapcraft.io/api/v1/snaps/sections"
```

If you do not know what might be causing this we will go to https://forum.snapcraft.io/ and ask there.

**Comment by pw10n** [comment]
hello. I believe I'm running in to the same issue here as well.

```
$ snap changes
ID   Status  Spawn               Ready  Summary
198  Doing   today at 16:59 UTC  -      Auto-refresh snap "microk8s"
$ snap tasks 198
Status  Spawn               Ready               Summary
Done    today at 16:59 UTC  today at 16:59 UTC  Ensure prerequisites for "microk8s" are available
Done    today at 16:59 UTC  today at 17:04 UTC  Download snap "microk8s" (2074) from channel "1.20/stable"
Done    today at 16:59 UTC  today at 17:04 UTC  Fetch and check assertions for snap "microk8s" (2074)
Done    today at 16:59 UTC  today at 17:04 UTC  Mount snap "microk8s" (2074)
Done    today at 16:59 UTC  today at 17:04 UTC  Run pre-refresh hook of "microk8s" snap if present
Done    today at 16:59 UTC  today at 17:06 UTC  Stop snap "microk8s" services
Done    today at 16:59 UTC  today at 17:06 UTC  Remove aliases for snap "microk8s"
Done    today at 16:59 UTC  today at 17:07 UTC  Make current revision for snap "microk8s" unavailable
Doing   today at 16:59 UTC  -                   Copy snap "microk8s" data
Do      today at 16:59 UTC  -                   Setup snap "microk8s" (2074) security profiles
Do      today at 16:59 UTC  -                   Make snap "microk8s" (2074) available to the system
Do      today at 16:59 UTC  -                   Automatically connect eligible plugs and slots of snap "microk8s"
Do      today at 16:59 UTC  -                   Set automatic aliases for snap "microk8s"
Do      today at 16:59 UTC  -                   Setup snap "microk8s" aliases
Do      today at 16:59 UTC  -                   Run post-refresh hook of "microk8s" snap if present
Do      today at 16:59 UTC  -                   Start snap "microk8s" (2074) services
Do      today at 16:59 UTC  -                   Remove data for snap "microk8s" (1910)
Do      today at 16:59 UTC  -                   Remove snap "microk8s" (1910) from the system
Do      today at 16:59 UTC  -                   Clean up "microk8s" (2074) install
Do      today at 16:59 UTC  -                   Run configure hook of "microk8s" snap if present
Do      today at 16:59 UTC  -                   Run health check of "microk8s" snap
Doing   today at 16:59 UTC  -                   Consider re-refresh of "microk8s"
```

It appears whenever snap decides to auto-refresh, microk8s hangs on the copy step and never completes (taking the cluster down).

The only things that seem to be effective were either rebooting the machine or recently discovered that I could abort the auto-refresh: 
```
$ sudo snap abort 198

... wait ...

$ sudo snap start microk8s
error: snap "microk8s" has "auto-refresh" change in progress

$ snap changes
ID   Status  Spawn               Ready  Summary
198  Abort   today at 16:59 UTC  -      Auto-refresh snap "microk8s"

$ snap list
Name      Version   Rev    Tracking         Publisher       Notes
core      16-2.49   10859  latest/stable    canonical✓      core
core18    20210128  1988   latest/stable    canonical✓      base
docker    19.03.13  796    latest/stable    canonical✓      -
helm3     3.1.2     5      latest/stable    terraform-snap  -
lxd       4.12      19766  latest/stable/…  canonical✓      -
microk8s  v1.20.2   2035   1.20/stable      canonical✓      disabled,classic
snapd     2.49      11107  latest/stable    canonical✓      snapd

$ sudo killall snapd
```


However, eventually the auto-refresh happens again...
```
~$ snap changes
ID   Status  Spawn               Ready               Summary
198  Undone  today at 16:59 UTC  today at 19:09 UTC  Auto-refresh snap "microk8s"
199  Doing   today at 19:14 UTC  -                   Auto-refresh snap "microk8s"
```

After reading this thread, I gave me the idea to look for unusual mounts that were lingering... and while i wasn't able to find references to libceph I did see that the nfs mounts from the nfs provisioner running in my cluster were erroring out. 

```
dmesg
[2465660.413978] nfs: server 10.152.183.19 not responding, timed out
[2465667.326132] nfs: server 10.152.183.19 not responding, timed out
[2465677.686357] nfs: server 10.152.183.19 not responding, timed out

mounts (partial)
10.152.183.19:/export/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/data on /var/snap/microk8s/common/var/lib/kubelet/pods/0af536f4-3349-4860-a73c-1c62ce6fede7/volume-subpaths/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/unifi/0 type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.5.0.10,local_lock=none,addr=10.152.183.19)
10.152.183.19:/export/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/log on /var/snap/microk8s/common/var/lib/kubelet/pods/0af536f4-3349-4860-a73c-1c62ce6fede7/volume-subpaths/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/unifi/1 type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.5.0.10,local_lock=none,addr=10.152.183.19)
10.152.183.19:/export/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/cert on /var/snap/microk8s/common/var/lib/kubelet/pods/0af536f4-3349-4860-a73c-1c62ce6fede7/volume-subpaths/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/unifi/2 type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.5.0.10,local_lock=none,addr=10.152.183.19)
10.152.183.19:/export/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/init.d on /var/snap/microk8s/common/var/lib/kubelet/pods/0af536f4-3349-4860-a73c-1c62ce6fede7/volume-subpaths/pvc-375dbf2c-5c54-4e64-a649-6059c8926017/unifi/3 type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=10.5.0.10,local_lock=none,addr=10.152.183.19
```

Not sure if this is the actual cause but thought I'd share in case it was helpful to anyone. Did anyone find a resolution to this problem?

**Comment by lpellegr** [comment]
I experienced the same issue. Pods remain in terminating states. New ones are created but fail to run due to a connectivity issue that I can't reproduce outside of the pod. Removing deployments and services before creating them does not help. I had to reinstall the whole cluster after a reset of all OS.

```
Name      Version   Rev    Tracking       Publisher   Notes
core      16-2.49   10859  latest/stable  canonical✓  core
core18    20210128  1988   latest/stable  canonical✓  base
lxd       4.0.5     19188  4.0/stable/…   canonical✓  -
microk8s  v1.20.4   2074   latest/stable  canonical✓  classic
snapd     2.49      11107  latest/stable  canonical✓  snapd
```

```
Mar 17 11:30:12 api-bhs snapd[47810]: stateengine.go:150: state ensure error: cannot sections: got unexpected HTTP status code 403 via GET to "https://api.sna>
Mar 17 11:30:24 api-bhs snapd[47810]: main.go:155: Exiting on terminated signal.
```

**Comment by t-o-o-m** [comment]
Came across the same problem - snap refreshed microk8s and took the cluster down.

How is this anywhere near a "reliable production-ready Kubernetes distribution" (from https://ubuntu.com/blog/introduction-to-microk8s-part-1-2) when 
1. one can't decide when updates should be performed (please don't mention the possibility to postpone or scheduling, I want to decide when to have potential downtimes)
2. a simple update takes the cluster down, with manual work needed to get it up and running again

I really like microk8s otherwise, but this renders it near useless.

**Comment by t-o-o-m** [comment]
Came across a probably related problem - snap refreshed microk8s and took the cluster down - all pods are in "sandbox changed" state then. Same goes for node reboots, btw. 
Microk8s restart usually helps, but sometimes I have to start from scratch to get it to work again.

1. one can't decide when updates should be performed (I know about the possibility to postpone or periodic scheduling, but I'd rather want to set a single point in time when to have potential downtimes). Might be possible with --devmode, but then I'd have to switch to a dev/edge channel
2. a simple update (and reboot in ny case) takes the cluster down, with manual work needed to get it up and running again

It would be great to tackle one of those two. Happy to provide any log, as I can easily reproduce the sandbox-issue.

I'd be careful with "reliable production-ready Kubernetes distribution" (from https://ubuntu.com/blog/introduction-to-microk8s-part-1-2) until then :)

**Comment by andrew-landsverk-win** [comment]
I also came across this issue today, also running rook-ceph in a 3 node cluster. The Rook/Ceph cluster works perfectly fine otherwise.

**Comment by lfdominguez** [comment]
Great, thanks snap autorefresh, you have crashed my entire cluster with this:
`/snap/microk8s/2338/bin/dqlite: symbol lookup error: /snap/microk8s/2338/bin/dqlite: undefined symbol: sqlite3_system_errno`

**Comment by pw10n** [comment]
Incase it's helpful to anyone here, I was able to permanently disable the auto-refresh by disabling the snapd service. 

`sudo systemctl stop snapd.service` then `sudo systemctl mask snapd.service` to disable
`sudo systemctl unmask snapd.service` and `sudo systemctl start snapd.service` to reenable

Since doing this, I haven't had any stability issues with my cluster at all. This is my temporary fix until I have time to migrate my cluster to k3s or something that actually works.

**Comment by ShadowJonathan** [comment]
These are all no permanent fixes, snap will never implement disabling auto updates, and this will always become a problem.

I just suggest not touching microk8s at all, only use it for development purposes, and ban it for all production purposes.

**Comment by ktsakalozos** [comment]
The Kubernetes project ships a few releases every month [1]. These releases include security, bug and regression fixes. Every production grade Kubernetes distribution should have a mechanism to release such fixes even before they are released from upstream. For MicroK8s this mechanism is the snaps. Snaps allow us to keep your Kubernetes infrastructure up to date not only with fresh Kubernetes binaries but also update/fix integrations with underlying system and the Kubernetes ecosystem.

If you do not want to take the risk of automated refreshes you have at least two options:

- use a snapstore proxy [3] to test revisions as they come and block those you do not want,
- set the snap refresh window in a convenient time for you [2]. This will not completely stop refreshes but will allow you to postpone them.


[1] https://github.com/kubernetes/kubernetes/releases
[2] https://docs.ubuntu.com/snap-store-proxy/en/
[3] https://snapcraft.io/docs/keeping-snaps-up-to-date

**Comment by ShadowJonathan** [comment]
@ktsakalozos the point of "security" is pretty moot if it breaks everything while updating it, it's defeating its own purpose.

**Comment by ktsakalozos** [comment]
@ShadowJonathan, I am not sure why you mention only security and in quotes. Any update that breaks the cluster is defeating its own purpose.

For anyone that wants to contribute back to this project we would be grateful if you could run non-production clusters with the candidate channels of the track you follow, for example `1.20/candidate`. Normally, the stable channel gets updated with what is on candidate after about a week. Having candidate releases well tested in a large variety of setup would be great.

**Comment by ShadowJonathan** [comment]
> I am not sure why you mention only security and in quotes.

Your point was that security is paramount and absolute, that it should be the excuse that makes this problem okay, it's not, it's an excuse that only exasperates this problem and the whole of snap for servers in general.

Snaps are fine for user apps, those can deal with being restarted, crashing, shutting down, again and again. Server apps need more delicacy, planning, and oversight. Any admin/operator would *not* want the developer control over when, how, and why something will update, they want complete control over their systems, and the snaps auto-updating feature is a complete insult to that.

> Any update that breaks the cluster is defeating its own purpose.

I'm glad you agree, then? I'd rather have a cluster which is outdated and vulnerable, and possibly get hacked, if it's about my own oversight and my own fault (at least then i can tune it to my own schedule and my own system). With auto-update, and even the update window, that control is taken away from me, as now i have to scramble to make sure the eventual update will not fuck with my system, and then to do it manually, safe, and controlled to make sure it does not fuck over the data. (which it did for me, 1.2TB of scraping data, all corrupted because docker didnt want to close within 30 seconds, after which it got SIGKILLd)

As a sysadmin, *I* control a developer's software, when, where, and how. The developer *doesn't* control my *system*, unless I *tell* it to. And even then, only on my own conditions.

Snaps violated this principle, and that's why I'm ***incredibly*** displeased with them.

**Comment by lfdominguez** [comment]
> @ktsakalozos the point of "security" is pretty moot if it breaks everything while updating it, it's defeating its own purpose.

i thinking same.... but they are telling that is production ready... really???

**Comment by ShadowJonathan** [comment]
@lfdominguez branding

**Comment by lfdominguez** [comment]
but if microk8s get out from snap.... or use another method, like executable self-contained (like k3s o k0s) i think that is better, you get out of the insane snap auto-refresh....

**Comment by lpellegr** [comment]
It looks like an option in _snapd_ to disable updates for a given package would satisfy most people who are complaining here (me including). 

Strangely, issues have been disabled on https://github.com/snapcore/snapd. I wonder where an issue could be created to discuss new options to snapd.

Unfortunately, it looks like people working on the project have a very strong opinion on the auto-refresh feature so I wonder if it is worth spending more time:
https://forum.snapcraft.io/t/disabling-automatic-refresh-for-snap-from-store/707/4

Perhaps, it's time to look at other solutions like [k3s](https://github.com/k3s-io/k3s) as it was already mentioned.

**Comment by lfdominguez** [comment]
> It looks like an option in _snapd_ to disable updates for a given package would satisfy most people who are complaining here (me including).
> 
> Strangely, issues have been disabled on https://github.com/snapcore/snapd. I wonder where an issue could be created to discuss new options to snapd.
> 
> Unfortunately, it looks like people working on the project have a very strong opinion on the auto-refresh feature so I wonder if it is worth spending more time:
> https://forum.snapcraft.io/t/disabling-automatic-refresh-for-snap-from-store/707/4
> 
> Perhaps, it's time to look at other solutions like [k3s](https://github.com/k3s-io/k3s) as it was already mentioned.

Disable the issues and call them itself as community driven.... is a shame, but well, mi cluster is already UP with 50 nodes, then i go with the `127.0.0.1` thing.... in a future if i get some free time i will fork snapd project and patch with an auto-refresh option....

**Comment by ktsakalozos** [comment]
> disable updates for a given package would satisfy most people who are complaining here

You are given this option. You can [block, schedule or postpone](https://github.com/ubuntu/microk8s/issues/1022#issuecomment-890531243) refreshes. You are even given the option to try an ["offline" manual deployment that never updates](https://github.com/ubuntu/microk8s/pull/1658/files). Although this may suit your needs, it is very hard from the Kubernetes distribution perspective to call this a good practice.

**Comment by lpellegr** [comment]
@ktsakalozos As it was explained, scheduling and postponing are not an option. We want full control, meaning we would like to disable auto-updates but keep the possibility to trigger an update manually when we want.

Blocking would be great but setting up a snapstore proxy looks overly complex and quite low level.

> You are even given the option to try an "offline" manual deployment that never updates.

Thanks for pointing this. This looks like a solution, although it seems we need to first download a snap file locally before installing it.

**Comment by lfdominguez** [comment]
Really, really... I have yet to read any real technical impediments that developers are relying on to not enable that functionality. All I read is whether it is better for the user or not, please not all users are dumb and there are many who manage technical services and not having that option is like telling them "You don't know what you are doing let us deal with it".

**Comment by ShadowJonathan** [comment]
> You are given this option. You can [block, schedule or postpone](https://github.com/ubuntu/microk8s/issues/1022#issuecomment-890531243) refreshes.

> Block

False, point me where exactly in your comment where you said that, and keep in mind I don't consider the snap store proxy to be a solution in this.

**Comment by lpellegr** [comment]
@lfdominguez I have the same feeling. Even if I understand the point of enabling auto-refresh by default, making everything possible and obscure to prevent people from disabling the feature is really sad.

**Comment by lalinsky** [comment]
And this happened again with the 1.19.14 release:

```
/snap/microk8s/2408/bin/dqlite: symbol lookup error: /snap/microk8s/2408/bin/dqlite: undefined symbol: sqlite3_system_errno
```

I was so happy with microk8s so far, but this is just bad.

**Comment by ktsakalozos** [comment]
@lalinsky where do you see this error? Is this on some daemon logs or do you get this when you execute a command?

Could you attach a microk8s.inspect tarball?

**Comment by JeongJuhyeon** [comment]
@ktsakalozos It's happening again now with 2483, same error as before 
`/snap/microk8s/2483/bin/dqlite: symbol lookup error: /snap/microk8s/2483/bin/dqlite: undefined symbol: sqlite3_system_errno`
 With the frequency that this has happened it's time for a big red warning sign on the main page saying "Do not use this for any multi-user service, auto-update is baked in and it may bring down your whole cluster".

**Comment by ktsakalozos** [comment]
@JeongJuhyeon could you please `sudo snap refresh microk8s` ?

**Comment by vazir** [comment]
Today a have experienced a crash of the PRODUCTION microk8s 3-nodes "HA" cluster. It just auto-updated to 1.21.5 ! As a programmer, admin, my mind even cannot comprehend what people deciding for the crucial services packaging have in mind to choose such a DNA broken tool as a snap??? Why at all UBUNTU uses it, when it hardly suitable even for desktop apps, and not suitable for services at all??? What is some medic.stuff would buy their adverting as "highly available" and people die because it auto-updates??? They must drop snap for anything aside the desktop apps, and better drop it at all and use proven by years .deb ...

**Comment by JeongJuhyeon** [comment]
@ktsakalozos It's one of the first things we tried but it did not help. Not sure about now, it's a staging server and after hours now.

@vazir Absolutely, we will now be migrating away from microk8s. It's straight up irresponsible that the docs/github page does not have a big warning saying "microk8s should not be used in any kind of situation where multiple people depend on it". It's obviously a complete deal-breaker for any kind of serious usage, going by this thread this has been an issue from day 1 and is never going to change with the unwillingness of snap to make autorefresh easy to disable. The only way any kind of auto-update system could possibly be used for production is if you're Microsoft and have complete control over every line of update code you ship as well as gigantic resources to test every possible setup and scenario. And even then they still let you disable it on systems meant for production environments.

**Comment by vazir** [comment]
I suspect people behind snap even scoff at us, as seems they are not allowing delaying auto-refresh more than for 60 days.
# snap set system refresh.hold=2050-01-01T15:04:05Z
# snap refresh --time
timer: 00:00~24:00/4
last: today at 10:22 UTC
hold: in 60 days, at 10:22 UTC
next: today at 17:25 UTC (but held)

**Comment by ktsakalozos** [comment]
@JeongJuhyeon please try to refresh.

@vazir, @JeongJuhyeon if you you want to take control over the updates you can either schedule them in a convenient time or install the snap store proxy mentioned in a previous comment. If you are certain you will not need updates in the future you can perform an offline deployment that will never update, we describe the few commands you need to run in the MicroK8s docs. You do however understand that MicroK8s cannot stop shipping patch releases and updates.

@vazir I am interested to know how did the cluster crash and what did you have to do to recover it? Thank you.

**Comment by ShadowJonathan** [comment]
@ktsakalozos i'm still awaiting an answer to [my question](https://github.com/ubuntu/microk8s/issues/1022#issuecomment-891188924)

**Comment by vazir** [comment]
> @vazir I am interested to know how did the cluster crash and what did you have to do to recover it? Thank you.

I had to restart all 3 nodes to bring it up again.

> you do however understand that MicroK8s cannot stop shipping patch releases and updates.

For the rest, you mentioned, you probably understand, that all you suggested are the hacks, and totally not suitable for a distribution? As nobody suggests "stopping patches distribution", but do you understand that NO ONE in the clear mind auto-updates critical services, like databases, kubernets clusters, and similar stuff? When you deploy a highly resilient, highly available, highly redundant stuff, which as k8s, which do on it's own ATTEMPTS to be resistant to failures, trying to keep itself alive... AND YOU DO kind of "kill -9" it from the outside??? What is the whole reason of developing microk8s at all than if you totally kill it's redundancy and resiliency by simply mischoosing the distribution tool?

**Comment by vazir** [comment]
@ktsakalozos I'm sure you do understand, that simultaneous update of the package on whatever number of the nodes, WILL lead to service stop, as ALL the nodes restarts simultaneously? For kubernetes, you (normally) do not restart the node, you drain it, making sure other nodes catch up for the pods, and than upgrade/maintain. So you just CANNOT distribute patches like you do. Yes, you just cannot. It may sound a litle strong, but may you imagine when lives depend on the stuff you do, or mass transport, or... whatever, where multiple persons affected? causing any service auto-restarts. You DO name your microk8s SUPER reliable, So ANY, even minor upgrades must be done by the admin, and admin might choose than to implement auto-upgrade, because the admin will do it right, making node drain, etc.

**Comment by Dart-Alex** [comment]
How is there still no option to disable auto updates and only do updates manually ? 
My production cluster went down today because snap killed a node. I had to work after hours to bring it back up. When ceph runs on a cluster, it needs very manual, and very gentle intervention to bring down a node gracefully. The way snap killed the node made connectivity errors pop up everywhere and I had to hard reboot every single node in the cluster to get back full control.
After a year and a half, it's unbelievable that we still can't disable auto updates !

**Comment by ktsakalozos** [comment]
Hi @Dart-Alex to completely stop snap refreshes you can do an offline deployment as described in [1], there is also the option to 
place a snap store proxy [2] in front of MicroK8s so you have better control on what revisions are applied and when.

Could you offer some more details on the errors you saw? Can I reproduce the issue? You mention the snap "killed a node". The expected behavior is that the node restarts, in your cases the node remained offline? If you are on 1.22 the workloads should remain untouched, was that the case for you? What steps did you follow to recover the system? Can you point me to any online  recovery instructions so I understand what happened? Thank you and apologies for the pain we may have caused.

[1] https://microk8s.io/docs/install-alternatives#heading--offline
[2] https://microk8s.io/docs/install-alternatives#heading--offline

**Comment by ktsakalozos** [comment]
> if I change channels to 1.22 after bringing down ceph on each node to allow snapd to bring them down, I should not have that problem in the future, even letting snapd auto refresh as it wants to ?

That sounds right based on what happened.

**Comment by Dart-Alex** [comment]
@ktsakalozos Also, if you want to reproduce, 3 nodes HA 1.21/stable with rook-ceph installed, osds on each nodes, some pods that use volumes from these osds, refresh one of the nodes and that should do the trick.

**Comment by ShadowJonathan** [comment]
@ktsakalozos i'm still awaiting an answer to [my question](https://github.com/ubuntu/microk8s/issues/1022#issuecomment-891188924), second time, and you are clearly active in this thread

**Comment by Dart-Alex** [comment]
@ShadowJonathan Same problem here, small startup, I'm the only sysadmin with limited resources, for now I have to resort to the "block all traffic to the official snap store until I have some time to manually handle updates" and I will have to keep on doing so even after updating to 1.22, just in case of an update breaking something unexpectedly. Snap in its current form seems like a really poor distribution medium for an all in one, easy to use HA cluster solution. The easy updates loose all their appeal once you learn that you are forced into them.

**Comment by jobh** [comment]
Based on what I've read here and elsewhere, I think that snap-the-philosophy is fundamentally unsound for high availability or production server deployments. When it comes to snap-the-technology, I find it darkly amusing that it has found a way of restarting k8s that corrupts my postgresql db. Which it now does reliably 4 times a day.

**Comment by joes** [comment]
Do you have any facts supporting your claim that microk8s restarts are continuously corrupting your postgresql db @johb? Is it a question of postgres going down due to restarts and writes failing during that time or is it something more insidious?

I also deployed postgres in our cluster and would like the microk8s team to have all the facts so that issues like this may be resolved.

I too have had an unfortunate incident with   snap-the-technology whereby microk8s HA-mode got disabled due to a snap refresh. The proposed fix was to move from a ”stable” channel to edge which seemed to fix the problem.

Understanding that the issue was caused by a refresh was not easy. IMHO this might have been made easier to discover if one was allowed to install a previous revision of a channel to compare with. Does one gain this ability if one enables the snap store proxy override feature?

**Comment by jobh** [comment]
@joes sorry for late reply. Nothing insidious, just that it gets corrupted by abrupt restarts when under write load. Why? I don't know, it's running at the bottom of a deep stack (VMWare, Kubernetes, XFS-on-LVM-through-hostpath-provisioner) and possibly one of those components handle fsync badly. I don't have the knowledge to figure it out. If I can do the restarts gracefully at my own schedule, there is no problem.

Now, I have tried various workarounds, all except store-proxy (way to complex, I don't need yet another infrastructure component to maintain and document):
1. Set the snap update schedule. It is not possible to set it to less than monthly though, also see next item.
2. Blackhole api.snapcraft.io in /etc/hosts. This worked fine for a while, but suddenly a few days ago snap began restarting microk8s 4 times a day (with associated corruption, constant restores). Even if there were no updates available, and even if I had set the schedule to monthly. This led me to write my earlier comment, which was perhaps too negative/sarcastic, sorry.
3. Stop and mask snapd in systemd. This seems to be successful so far, but it's a really big hammer for such a simple problem.

**Comment by jobh** [comment]
Btw; you asked for facts, I have only circumstantial evidence. My "snap changes" log started getting lines like this every four hours,

```172 Done today at 04:51 CEST today at 04:51 CEST Running service command for snap "microk8s"```

and several of these matched the known time of postgresql restart and also start-of-corruption. (this specific line is quoted from an early comment in this thread, I can't see my own unless I re-enable snapd, which I don't dare to do outside of maintenance window)

**Comment by joes** [comment]
Meanwhile an automatic snap refresh broke one node in my cluster again (see #2831). I guess I will have to try your solution 3 @johb. And figure out how to unmask (?) for those times I have a need to refresh (after testing on another environment).

**Comment by jobh** [comment]
...but given that snap is what it is, and microk8s is what it is, I'll try to bring this back to something constructive.

After some experimentation, it looks like the basic problem is that `microk8s.stop` brings down the k8s infrastructure (including the hostpath provisioner) before the containers have had a chance to react to the SIGTERM and finish gracefully. Would it be possible to change either microk8s.stop or the snap refresh hook so that the containers are allowed to terminate before the infrastructure disappears?

This is my test script, which I piped to a hostpath-provisioned folder. When bringing the container down via kubectl, it counts to 30 seconds. When doing `microk8s.stop`, I see nothing after "Waiting for SIGTERM", even though I can see the process live for 30 seconds after that (in `ps`).
```
import os
import signal
import time

def handle_sigterm(*args):
    print('Got SIGTERM')
    i = 0
    while True:
        time.sleep(1)
        i += 1
        print(f'{i} seconds since SIGTERM')

if __name__ == '__main__':
    print(f'PID {os.getpid()}')
    signal.signal(signal.SIGTERM, handle_sigterm)
    print('Waiting for SIGTERM')
    while True:
        time.sleep(10)
```

**Comment by ShadowJonathan** [comment]
I would like to mention that this is likely the very same problem that corrupted my database as well, killing or leaving containers in an orphaned and errored state will very likely corrupt *any* stateful application.

Snapcraft does not give tools, nor really warns administrators, to counter this sort of behaviour. Snap's refresh functionality fundamentally makes server applications unstable in the long term, and the level of quality it assures and pumps out for these applications is far lower than a mechanism like refresh would require, it is far to easy to have a developer inadvertently corrupt every user's data.

**Comment by jobh** [comment]
Yup. I guess there are really two issues at play here:
1. A reasonable way to disable auto-updates, which are harmful in some circumstances. Controversial, may not happen.
2. Fix corruption of stateful containers on `microk8s.stop`. Uncontroversial (probably), will make # 1 less harmful, also helps e.g. non-snap-triggered restarts and reboots.

Should # 2 get a separate issue?

**Comment by a-hahn** [comment]
Interesting read from a former employee of canonical working in the snap advocacy team on howto disable auto-updates efficiently: https://popey.com/blog/2021/05/disabling-snap-autorefresh/

**Comment by ktsakalozos** [comment]
Hi @a-hahn, some of the options you have to handle autorefreshes are discussed in our docs. You may want to look at [1].

[1] https://microk8s.io/docs/snap-refreshes

**Comment by a-hahn** [comment]
@ktsakalozos Thx ! I have updated related https://discuss.kubernetes.io/t/snap-refreshes/18791 with this important note on how to disable snap auto-refresh easily and completely.

**Comment by ktsakalozos** [comment]
Thank you for this @a-hahn. The section you added describes the process linked to in `Setup a store disconnected deployment`.

**Comment by ktsakalozos** [comment]
@a-hahn it is hard for us to recommend in the official docs that deployments could completely stop updates. This 'breaks' the intended behavior of snaps. I appreciate your understanding in not accepting the docs update.

**Comment by evilnick** [comment]
just to add, the reasonable way to gate updates is to use a snap store proxy, as also covered in the docs. This doesn't 'break' the install by preventing any updates, but does make it possible to stop updates happening automatically (or at all if you wish)

**Comment by a-hahn** [comment]
@ktsakalozos well i understand that its 'hard to recommend' but I think it is well worth mentioning because its an obvious and simple solution for many users. Trying to hide that piece of information just hurts your / Canonicals reputation as a first class solution provider. This is even more true as snaps often stack upon each other (when you consider microk8s installed in lxc containers both installed via snaps) making troubleshooting much more difficult as it should be. Even more I can't see anything positive about auto-updates as they are a backdoor into the installation and who knows which hidden features or paywalls might be introduced in the future. Misguiding people never pays in the long run. I think you should think twice about not accepting the docs update. I am alarmed by now.

**Comment by ktsakalozos** [comment]
@a-hahn the intended use of MicroK8s is to run it with updates on. In the documentation of the product we want to recommend the intended use.

>  I can't see anything positive about auto-updates

The issue under question here is why the intended use is to run Kubernetes with updates on. I hope this will become clear with an example. K8s v1.23.0 came out on the 7th of Dec, v1.23.1 on the 16th of Dec, v1.23.2 on the 19th of January, v1.23.3 on the 26th of January [2]. In parallel we have updates on  underlying components such as containerd, runc etc. You can check what some of these updates are about in [1] and after reviewing them you may say that you are not interested in them and that is absolutely fine. But on our side, the side of a K8s distribution, we cannot just ignore updates. We will continue providing updates and we would like these updates to be applied to already existing deployments because we strongly believe if they are not the end product will degrade in quality. It would have been irresponsible for us to ship a product knowing that within 15 days would be out of date and thus would hurt Canonical's reputation as a first class solution provider much more. In the same way you allow updates on your OS, your phone, your tablet you are expected (intended use) to allow updates on this K8s distribution. I hope this clarifies why we take this position in respect to updates and why we are not "misguiding people".

For the case where the user wants to have full control over what updates get applied and when the recommendation is to not disconnect the deployment from updates but to use a proxy to filter-out and block updates that you deem harmful. This process is described in our official docs [3], and, yes, it is not as easy as download the snap binary and install it but it is a process that makes you aware of updates.

After all this, if you still want to download the snap binary and install it the way you suggest the snap ecosystem will not stop you. You are given the `--dangerous` flag to serve this purpose. Yet, once more, this is not the recommended way to use MicroK8s and therefore we would not promote it in our official docs. A simple google search will lead you down that path.

Also, please allow me this last point. I find the "paywall" comment unfair for the company that has been giving you all these years Ubuntu, cloud-init, Launchpad, MaaS, LXD, juju OpenStack, Multipass, MicroK8s, Charmed K8s, snapcraft and more.

[1] https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.23.md
[2] https://github.com/kubernetes/kubernetes/releases?page=1
[3] https://microk8s.io/docs/manage-upgrades-with-a-snap-store-proxy

**Comment by evilnick** [comment]
@a-hahn Hi. In respect of the documentation, no matter how many warnings or comments are added to things, sadly many people don't read them. There is no hidden agenda, this is simply a matter of responsibility. If you buy a car and want to disconnect all the warning lights, that is also up to you, but you wouldn't expect to find the instructions to do so printed in the owners manual. Please add your method as a comment on discourse if you like.

**Comment by ShadowJonathan** [comment]
> Please add your method as a comment on discourse if you like.

That'll be nothing more than black-holing the concern.

**Comment by evilnick** [comment]
> > Please add your method as a comment on discourse if you like.
> 
> That'll be nothing more than black-holing the concern.

What is the actual concern?

**Comment by a-hahn** [comment]
@ktsakalozos So much time spend on arguing. This simple one-liner `snap download microk8s && snap install microk8s --dangerous` is a well-known alternative install for the interns at canonical as this  [blog post proves](https://popey.com/blog/2021/05/disabling-snap-autorefresh/). It would have saved lots of hours of troubleshooting for the people contributing to this post including @jobh @ShadowJonathan @Dart-Alex @joes @skobow @eug48 Thx again @ktsakalozos @evilnick for your helpful support on this topic.

**Comment by evildmp** [comment]
@a-hahn Hello Andreas, thank you for your engagement on this, and for making a contribution to the documentation for this project.

I'm a Director of Engineering at Canonical, and my responsibility is documentation - everything to do with documentation and the way we produce it, for all Canonical products and projects (if you'd like to learn more about what my work is, I invite you to read [The future of documentation at Canonical](https://ubuntu.com/blog/the-future-of-documentation-at-canonical)).

Needless to say, I think that documentation is extremely important. I think that's generally true, but particularly so for an open-source software organisation like Canonical, and for open-source software projects themselves. That's because documentation is part of the contract with an open-source community. Documentation is one of the most important ways of sealing the relationship between a product and its community. 

Community members understandably feel strongly about documentation. You put time and effort into making an improvement to the MicroK8s documentation, and it was declined - for reasons that you don't agree with - by one of the maintainers. I can see that you are angry and frustrated that the contribution you made was reversed, and your reasoning about it also not accepted.

Can I ask, are you upset because you think @evilnick made a wrong technical decision about what should or should not be documented? Or would you say you feel more upset because you need to feel a different relationship to exist between you (as a community member), and the project, and Canonical?

Unfortunately I am not in a position to comment on the technical aspect of this. I am not an expert in MicroK8s security. However, one of the things I would like to achieve in my work is improved community engagement through documentation and improved experiences for documentation contributors, so I'd be very happy if you would like to talk more about that.

Either way, one thing I would like to say in the meantime is that I do know that Nick is a very community-minded person. Before making this decision, we discussed it together (nor was I the only person it was raised with), because it's a hard thing to do to turn down someone's contribution. It was not done lightly. So it's also a hard thing to be in that position, and to receive angry criticism for it, or to be accused of not respecting the code of conduct. Personally, I would be upset by that myself.

**Comment by ShadowJonathan** [comment]
@evildmp if I were to guess, a large share of the frustration is not personal, but rather aimed at snap's packaging in-of-itself, which is then a main root to cause this problem.

The solution offered is a hack, an explicit circumvention of the problem, which does not do much to offer a satisfying resolution, nor does it help lighten the burden that the problem caused, it only cripples the effectiveness of the platform, while a better solution is available from snap's side, while they do not wish to give developers those tools, out of political and ideological reasons, that explicitly tear away control from users in a patriarchal fashion.

I don't want to perpetuate the cycle here, at the very least know this; it wasn't personal, the frustration is high, and this issue is just one part of the knot where the pressure became too high.

**Comment by vazir** [comment]
The approach of the mikrok8s team is not professional. The sole purpose of kubernetes is to provide a platform for running fault tolerant services, which is completely ruined by packaging it into totally unsuitable, and DNA broken "snap" tool. It's hard to say more. I personaly got rid of all microk8s on my servers and migrated to k3s. Next is to replace back ubuntu by plain Debian.

**Comment by evildmp** [comment]
@vazir 

> The approach of the mikrok8s team is not professional.

This is not an issue, or criticism of code, or even criticism of someone's behaviour. It is an abusive remark. There is nothing constructive that anyone can do in response to this comment.

Sometimes people get angry about open-source software projects they participating, which is OK. It's also OK to express anger sometimes.

It is not OK, and it is explicitly against [the Ubuntu code of conduct adopted by the MicroK8s project](https://ubuntu.com/community/code-of-conduct), to make abusive comments.

I politely request that you delete that part of your comment. Thank you.

**Comment by lfdominguez** [comment]
I think that some last comments is off-topic related to title of the issue..... We have a problem... the autorefresh system of snap breaks microk8s, we need an option to disable autorefresh... only that...

**Comment by vazir** [comment]
@evildmp , 
It has nothing to do with opensource and etc. I do not criticize the code, but the wrong packaging way itself, which breaking WHOLE idea of having microk8s, and i'm using straight words for that. Once again I will try to explain what is wrong, it's of course useless, but anyways. Looking at the meaning of the messages of other user of this thread: everyone tells the same, words different. Everyone says "It's broken, it's unacceptable, change it, do something". If there is no other way to multiplatform distiribution of microk8s, just state in the docs, in front, very visible and understandable "DO NOT USE IT FOR PRODUCTION WHEN INSTALLED VIA SNAP BECAUSE YOUR SERVICE MAY BREAK RANDOMLY, UNEXPECTEDLY". And this way, no one will have any issues. We will just look for other k8s from the start, and use microk8s (probably, not really) for dev only, and not experience the problems randomly and unexpectedly in production, because, according to the front page it's rock solid reliable. Stating in the docs somewhere deep deep "install manually if you do not like snap", changes nothing, just making users just more disappointed.


> @vazir
> ...

**Comment by lpellegr** [comment]
Said differently, why is there so much reluctance to improve user experience with a feature so many people requests? Is there something to do with data collected from auto updates?

**Comment by ShadowJonathan** [comment]
@vazir id also like to note that the current Ubuntu server installer, when entering the "additional software" screen, installs these through snap.

So, if you've installed docker on that screen, it'll go ahead and install that in a snap container, including any other software, such as microk8s.

**Comment by vazir** [comment]
@ShadowJonathan - this mindless practice effectively and rapidly moves Ubuntu as distribution out of servers. I do not believe, they do not understand it. So, there is the only conclusion - someone slowly killing Ubuntu from inside. Like Nokia.

**Comment by neoaggelos** [comment]
A note that is probably useful for people that have commented in the issue:

Starting with snapd 2.58, it is possible to indefinitely hold MicroK8s (and any other installed snap packages) from updating with the following command:

```
sudo snap refresh --hold microk8s
```

See also the "Control updates" section in the [snapd documentation](https://snapcraft.io/docs/keeping-snaps-up-to-date)

Not stale

**Comment by vazir** [comment]
> 

I can only say "finally"... But I dropped microk8s, switched to k3s, also no UBUNTU any more. Back to debian.

**Comment by yuryskaletskiy** [comment]
Had switched to Rancher/k3s already. No further bothering on unexpected auto updates.

**Comment by JWilson45** [comment]
this just happened to me, failed on 2/3 nodes

**Comment by fmiqbal** [comment]
gosh darn it, i am now really suspecting snap auto refresh is why my cluster is somewhat unstable in specific interval, today snap auto refresh to 1.29.11, 3 of 9 worker nodes didnt restart properly, 
![image](https://github.com/user-attachments/assets/5cfea7a6-c60c-4ca4-aff7-768c81782a19)

the fix is easy, stop and start microk8s, why !?

Moreover, because we use longhorn, when nodes become "NotReady" as stated by API Server, the volume will be rebalanced automatically, which make a lot of request to API Server and make it hard to query (i hate you dqlite), for some time we use 16 GB x 3 for master node, and when this thing happens (I suppose the update), the cluster can become totally down, nothing is running, dqlite is choked, the kine count (`sudo -E /snap/microk8s/current/bin/dqlite -c /var/snap/microk8s/current/var/kubernetes/backend/cluster.crt -k /var/snap/microk8s/current/var/kubernetes/backend/cluster.key -s file:///var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml k8s "select count(*) from kine`) is growing fast, with negative feedback loop to everything in the cluster and in matter of hours it can bring whole cluster down, need to restore from "cold boot"

this time around we suspect 16GB is not enough and increase it to 32GB, at least the kine count can be manageable (albeit still getting bigger over time, but a lot longer than before), and we restore (literally just restart the nodes) in time before everything goes to sh*t for the nth time over 6 month period

**Comment by gtupak** [comment]
> A note that is probably useful for people that have commented in the issue:
> 
> Starting with snapd 2.58, it is possible to indefinitely hold MicroK8s (and any other installed snap packages) from updating with the following command:
> 
> ```
> sudo snap refresh --hold microk8s
> ```
> 
> See also the "Control updates" section in the [snapd documentation](https://snapcraft.io/docs/keeping-snaps-up-to-date)
> 
> Not stale

@neoaggelos even if you do this, isn't it also possible that if snap updates any other package it can cause pods to restart? I've experienced this in my environment, where snap updated `multipass` and caused all pods to restart (even though ironically enough I wasn't even using `multipass` on the node)

**Comment by miks** [comment]
My single-node microk8s started failing twice a day out of nowhere. I can fix broken pods stuck in `CrashLoopBackOff` by manually stopping and starting microk8s.
After looking for possible culprits, it seems this is the issue:
```
# journalctl -b -u snapd.service
Oct 15 01:51:51 k8s-dev.cube.lv snapd[989]: storehelpers.go:916: cannot refresh: snap has no updates available: "core20", "core22", "snapd"
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-apiserver-kicker.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-etcd.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-kubelite.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-flanneld.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-containerd.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-apiserver-proxy.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-k8s-dqlite.service
Oct 15 01:51:58 k8s-dev.cube.lv snapd[989]: services.go:1152: RemoveSnapServices - disabling snap.microk8s.daemon-cluster-agent.service
Oct 15 01:52:09 k8s-dev.cube.lv snapd[989]: storehelpers.go:916: cannot refresh snap "microk8s": snap has no updates available
```

Each time this happens, many pods enter the `CrashLoopBackOff` state. There have been no updates to the OS, no major version changes to microk8s, and nothing else.
It simply started crashing itself twice a day.
Is this by design?

**Comment by JWilson45** [comment]
@miks I switched to k3s and have had a much better experience.

**Comment by miks** [comment]
> [@miks](https://github.com/miks) I switched to k3s and have had a much better experience.

That's my plan for now as well. I can't allow a system to break down on its own without external issues like low disk space or memory problems.

**Comment by HomayoonAlimohammadi** [comment]
Hello @miks,

Thank you for raising this issue and I'm really sorry for the inconvenience. I can't be sure about what's happening in your cluster, but it seems like snapd is disabling Microk8s services before a refresh that never actually happens (since there's no new version). Regardless, you can hold the snap to prevent it from getting refreshed automatically:
```
snap refresh --hold=<duration> microk8s
```
Please refer to the docs for more details: https://snapcraft.io/docs/managing-updates
